### ISSUE TITLE
Spring integration

### ISSUE THREAD

**Issue by slinkydeveloper** [cloudevents/sdk-java#142]
Given the ongoing work on [jaxrs implementation](https://github.com/cloudevents/sdk-java/issues/121), what it takes to integrate with spring?

**Comment by slinkydeveloper** [comment]
Ping @bsideup

**Comment by bsideup** [comment]
cc @olegz (Spring Cloud)

**Comment by olegz** [comment]
I guess there should be clarification on what type of integration you are looking for. Spring provides variety of ways to expose user functionality as REST endpoints. Some of them require user to no configuration at all. I can certainly elaborate further but it would help if more specifics are shared

**Comment by slinkydeveloper** [comment]
The Jax-Rs integration will consists of a `@Provider` implementing [`MessageBodyReader`](https://javaee.github.io/javaee-spec/javadocs/javax/ws/rs/ext/MessageBodyReader.html) and [`MessageBodyWriter`](https://javaee.github.io/javaee-spec/javadocs/javax/ws/rs/ext/MessageBodyWriter.html)

**Comment by dsyer** [comment]
There are many equivalents to `MessageBodyReader` and `MessageBodyWriter` in Spring (depending on blocking or non-blocking, HTTP, AMQP, Kafka, JMS, etc.). I think rather than try and implement them all here it would be better just to present a clean, stable API.

**Comment by slinkydeveloper** [comment]
@olegz @dsyer draft of the jaxrs implementation is here: https://github.com/cloudevents/sdk-java/pull/145

Can it work OOTB with Spring?

**Comment by dsyer** [comment]
Anyone who used JAX-RS with Spring I imagine could register the `@Provider` - I think that's the way it works with Spring Boot. The `MessageUtils` looks useful as well for internal use in Spring maybe (I see you used it in the Kafka binding already).

**Comment by slinkydeveloper** [comment]
I've merged the jax-rs integration https://github.com/cloudevents/sdk-java/tree/master/http/restful-ws

Let me know if it can work OOTB with Spring or if we need something else :smile:

**Comment by bsideup** [comment]
@slinkydeveloper when you're saying "work OOTB with Spring" - do you mean that this is enough to call it "Spring support", or that, when Spring's JAX-RS integration is used, the module works?
Because Spring is not _based_ on JAX-RS (but there are options to _use_ JAX-RS with Spring)

**Comment by slinkydeveloper** [comment]
> Because Spring is not based on JAX-RS (but there are options to use JAX-RS with Spring)

Ah ok, please forget my ignorance about that :smile:

**Comment by cmoulliard** [comment]
For Spring, then we should add a new module within this SDK implementing `Spring Web`- https://github.com/spring-projects/spring-framework/tree/master/spring-web

**Comment by slinkydeveloper** [comment]
@cmoulliard are you willing to give a try to it? :smile:

**Comment by bsideup** [comment]
@slinkydeveloper I would highly recommend to contact the Spring team before proceeding, to ensure that what you're doing here is aligned with what will be the official CloudEvents support in Spring.

**Comment by cmoulliard** [comment]
> what will be the official CloudEvents support in Spring.

Where such official CloudEvents vision is defined and communicated for the spring users/developers ? @bsideup

**Comment by olegz** [comment]
@cmoulliard at the moment it is not very clear as to what does _Spring support_ even mean, since Spring portfolio consists of many projects. As for _communication_ . .  as an example, we have request in [spring-cloud-function](https://github.com/spring-cloud/spring-cloud-function/issues/422) to add Cloud Events support which most likely be a simply translation layer between Spring [Message](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/messaging/Message.html) and [CloudEvent](https://github.com/cloudevents/sdk-java/blob/master/api/src/main/java/io/cloudevents/CloudEvent.java), since their structures are pretty much identical. 

Basically IMHO I don't think the terminology such as _"CloudEvents support in Spring"_ or _"Spring Integration" - (which is also the name of one of Spring's frameworks)_ is appropriate or meaningful without further clarifying which specific part of Spring one is referring to.

**Comment by cmoulliard** [comment]
> support

I used the word `vision` and not `support` at all ;-) @olegz

**Comment by slinkydeveloper** [comment]
>  As for communication . . as an example, we have request in spring-cloud-function to add Cloud Events support which most likely be a simply translation layer between Spring Message and CloudEvent, since their structures are pretty much identical.

Is there any reason why we can't have this integration inside this community project (similar to how we've done for kafka, jax-rs and vertx)? It sounds a pretty lightweight integration that can live under some `cloudevents-spring` module or similar

**Comment by olegz** [comment]
@cmoulliard i was referring to the "quoted text" - `what will be the official CloudEvents support in Spring.`

**Comment by salaboy** [comment]
@cmoulliard @slinkydeveloper I've just created a PR with some Spring Boot (Web and WebFlux) Helpers -> https://github.com/cloudevents/sdk-java/pull/201
I will be happy to send that to Spring Boot / Spring Cloud -> @spencergibb @cmoulliard @olegz any idea where?

**Comment by olegz** [comment]
As it was suggested by Dave [earlier](https://github.com/cloudevents/sdk-java/issues/142#issuecomment-620570544), there are many types of abstractions in spring representing different types of readers and writers specific to the context of current execution etc. And then of course there are many projects where CloudEvents may be relevant, so I am not sure how feasible it is to manage all of them in isolation from where they are most relevant to be used.
Also, as it was mentioned before, CloudEvent is pretty much identical to Spring [Message](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/messaging/Message.html); All Spring projects understand Spring [Message](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/messaging/Message.html), so all we really need to do is to provide implementation of [MessageConverter](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/messaging/converter/MessageConverter.html) to convert CloudEvent to/from [Message](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/messaging/Message.html). Basically instead of making every Spring bit understand CloudEvent which may result in much greater effort with a lot of duplication, we provide an implementation of MessageConverter which Spring will pick up automatically and user would not even have to know that they are dealing with CloudEvents. In fact we've had the same approach for [Message](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/messaging/Message.html) making it more of an option for user to use while it is still used internally.

**Comment by salaboy** [comment]
@olegz that makes a lot of sense to me .. I've provide a simple helper to get people started.. we can change the implementation to use Spring Message underneath later. I want to make sure that we unblock spring users to start using Cloud Events asap.. Do you think that is worth using Spring Message even for these simple helpers?

**Comment by olegz** [comment]
@salaboy I understand what you're trying to do and unfortunately I don't have a clear answer at the moment. That is perhaps the reason for Dave's earlier comment about simply "ensuring stable API":

> There are many equivalents to `MessageBodyReader` and `MessageBodyWriter` in Spring (depending on blocking or non-blocking, HTTP, AMQP, Kafka, JMS, etc.). I think rather than try and implement them all here it would be better just to present a clean, stable API.

May be it is better to to let individual spring projects to handle it. As I mentioned before we have an open issue in [spring-cloud-function](https://github.com/spring-cloud/spring-cloud-function/issues/422) which most likely will result in the implementation of the MessageConverter and then the question would be to push/migrate such implementation up to Spring Messaging so every projects can benefit from it automatically without user doing anything at all.

**Comment by cmoulliard** [comment]
> As I mentioned before we have an open issue in [spring-cloud-function](https://github.com/spring-cloud/spring-cloud-function/issues/422) which most likely will result in the implementation of the MessageConverter

Such `MessageConverter implementation` should not be tied to the function project as Cloud Events will be used from web, webflux applications too

**Comment by rstoyanchev** [comment]
Looking at PR #201, this is not an idiomatic way to integrate. I would suggest implementing an `HttpMessageConverter<CloudEvent>` instead and then it becomes:
```java
    @PostMapping
    public String recieveCloudEvent(CloudEvent cloudEvent) {
        // ...
    }
```

Or:
```java
CloudEvent cloudEvent = ... 
restTemplate.postForEntity(url, cloudEvent, String.class)
```

For WebFlux, it would be a pair of `Encoder<CloudEvent>` and `Decoder<CloudEvent>` implementations that would allow the same server side signature and and this with WebClient:

```java
CloudEvent cloudEvent = ...
webClient.post().uri(url).bodyValue(cloudEvent).retrieve()
```

**Comment by olegz** [comment]
@cmoulliard 
The role function project plays in the context of this thread is only as the origin of the issue to support CloudEvents in Spring. 
The `MessageConverter` approach is in no way tied to function or any other Spring project and rather a general purpose solution to provide ability to turn `CloudEvent` into Spring `Message` and vice versa. And of course as @rstoyanchev  suggested the `HttpMessageConverter`, `Encoder` and `Decoder` would play the same role for various different style of web request.
